stdlib/event/gui.lua

```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
+        if reg_index and reg_index ~= #Gui._registry[event_ids][gui_element_pattern] then
+            table.remove(Gui._registry[event_ids][gui_element_pattern], reg_index)
+            log(warnmsg ..  "reordering to bottom.")
+            table.insert(Gui._registry[event_ids][gui_element_pattern], handler)
+            Event.register(event_ids, Gui.dispatch)
+            return Gui
+        elseif reg_index then
+            log(warnmsg .. "re-registering event.")
+            Event.register(event_ids, Gui.dispatch)
+            return Gui
+        end
+    end
+
+    table.insert(Gui._registry[event_ids][gui_element_pattern], handler)
+    return Gui
+end
+
+--- Removes the handler with matching gui element pattern from the event.
+-- @tparam defines.events event_ids valid values are `defines.events.on_gui_*` from @{defines.events}, or
+-- a list of them
+-- @tparam string gui_element_pattern (optional) name or string regular expression for a handler to remove
+-- @tparam function handler (optional) handler to deregister from specified Gui events.
+-- @return (<span class="types">@{Gui}</span>)
+function Gui.remove(event_ids, gui_element_pattern, handler)
+    fail_if_missing(event_ids, "missing event argument")
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.remove(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
+
+    if gui_element_pattern and type(gui_element_pattern) ~= "string" then
+        error("gui_element_pattern argument must be a string")
+    end
+
+    if not Gui._registry[event_ids] then
+        log("Request to remove non-registered gui event handler for " .. event_ids .. " ignored.")
+        return Gui
+    end
+
+    if gui_element_pattern and not Gui._registry[event_ids][gui_element_pattern] then
+        log("Request to remove non-registered <event, pattern> tuple: <" .. event_ids ..
+            ", " .. gui_element_pattern .. "> ignored.")
+        return Gui
+    end
+
+    -- recurse, if needed, to replace the optional arguments
+    -- with concrete values from the registry.
+    if not gui_element_pattern then
+        for pattern, _ in pairs(Gui._registry[event_ids]) do
+            Gui.remove(event_ids, pattern, handler)
+        end
+        return Gui
+    elseif not handler then
+        local reg = Gui._registry[event_ids][gui_element_pattern]
+        for i = #reg, 1, -1 do
+            Gui.remove(event_ids, gui_element_pattern, reg[i])
         end
-        if table.size(Event._registry[event_id]) == 0 then
-            Event._registry[event_id] = nil
-            script.on_event(event_id, nil)
+        return Gui
+    end
+
+    local reg = Gui._registry[event_ids][gui_element_pattern]
+    for i = #reg, 1, -1 do
+        if reg[i] == handler then
+            table.remove(reg, i)
         end
     end
+    if not table.any(reg, function (v) return true; end) then -- luacheck: ignore v
```

Comment:
`if table.size(reg) == 0 then` is still a much faster check, PS  for the future function() instead of function(v) since you are not using any parameters

Reply:
>>`table.any(reg, function(v) return true; end)`
>
>`if table.size(reg) == 0 then` is still a much faster check

I am a bit mixed up about table internals, different versions of lua, etc...  For now I'm just leaving it but I do intend to reevaluate after my next visit to the clue depot.  Maybe we should add something like this to stdlib/event/table?:
```
function table.empty(t)
  return next(t)==nil
end
```

> function() instead of function(v) since you are not using any parameters

Lua will just let a function's consumer throw arbitrary parameters into a black hole like that?  *shudder*... Would luacheck be happy about it?

Reply:
Well it isn't really a black hole  function(v) creates a local variable v and puts whatever the function passes into it,  if you are not using v at all no point even adding it to the parameters just to ignore it. and luacheck will be happy.

table_size is a factorio function which returns a c++ count of keys which is faster than calling next(t), table.size is just a re-mapping to add it to the table mutates, stdlib also accounts for the potential lack of table_size (ie for busted tests) by treating table.size as table.count_keys if table_size is nil.

If it is not already present I will add table.isempty and Is.Empty, as well as table.clear

Reply:
Nexela said:
> table_size is a factorio function which returns a c++ count of keys which is faster than calling next(t), table.size is just a re-mapping to add it to the table mutates, stdlib also accounts for the potential lack of table_size (ie for busted tests) by treating table.size as table.count_keys if table_size is nil.

That explains everything.  I thought we were talking about a lua intrinsic, but I must have been thinking of getn.

Now that I think about it, in working on this stuff, I may have discovered that next is broken for certain factorio userdata things.  For example, I am pretty sure that next(game.players, next(game.players)) == "__self" or something crazy like that.  My conclusion was: if x,y,z = pairs(game.players), somehow, x ~= next.   I could be mixed up; regardless, I think something tricky is happening when you iterate over game.players... anyhow I digress.

Reply:
> If it is not already present I will add table.isempty and Is.Empty, as well as table.clear

That would be fantastic, thanks/tia (depending on whether or not you already did so, respectively :))

Reply:
I wrote:
> Now that I think about it, in working on this stuff, I may have discovered that next is broken for certain factorio userdata things. For example, I am pretty sure that next(game.players, next(game.players)) == "__self" or something crazy like that. My conclusion was: if x,y,z = pairs(game.players), somehow, x ~= next. I could be mixed up; regardless, I think something tricky is happening when you iterate over game.players... anyhow I digress.

But, I looked again, and, in a single-player game,

serpent.line({next(game.players)}) => {"__self", "userdata..."}
serpent.line({next(game.players, next(game.players))}) => {}

So, it seems, game.players is a completely normal table containing an array of player userdata objects, precisely as I would expect it to be.  Instead the real problem appears to be a pebcak, specifically, my tendency, (amusingly identifiable as a bug in the code quoted above) to sometimes forget that next is returning two values rather than a two-element array.